### PR TITLE
Optimized Routing added through Mapzen

### PR DESCRIPTION
**Pull request by lorilith**

### CODE REVIEW

File: PoGo.PokeMobBot.Logic/Routing.cs [hunk 1]
```diff
@@ -47,7 +48,69 @@ public static RoutingResponse GetRoute(GeoCoordinate start, GeoCoordinate dest)
...
+                url_req = "https://matrix.mapzen.com/optimized_route?json={\"locations\":[";
+                url_req += $"{{\"lat\":{session.Client.CurrentLatitude},\"lon\":{session.Client.CurrentLongitude}}}";
+                foreach (FortCacheItem item in list)
+                {
+                    url_req += $",{{\"lat\":{item.Latitude},\"lon\":{item.Longitude}}}";
+                }
+                url_req += $"],\"costing\":\"pedestrian\",\"units\":\"km\"}}&api_key={matrixkey}";
+                //Logger.Write($"JSON:\n{url_req}");
+                WebRequest request = WebRequest.Create(url_req);
+                request.Credentials = CredentialCache.DefaultCredentials;
```
Comment (bmdr38) [inline PoGo.PokeMobBot.Logic/Routing.cs]: Are credentials actually required?  I have my doubts

File: PoGo.PokeMobBot.Logic/Routing.cs [hunk 2]
```diff
@@ -47,7 +48,69 @@ public static RoutingResponse GetRoute(GeoCoordinate start, GeoCoordinate dest)
...
+                    List<FortCacheItem> tmpList = list.Where(i => Math.Abs(i.Latitude - coord.ToArray()[1]) < 0.00002 && Math.Abs(i.Longitude - coord.ToArray()[0]) < 0.00002).ToList();
+                    if (tmpList.Count > 0)
+                    {
+                        ret_list.Add(tmpList.ToArray()[0]);
+                    }
+                }
+                Logger.Write("Routing Order processed.", LogLevel.Debug);
+                return ret_list;
+            }
+            catch (Exception ex)
```
Comment (bmdr38) [inline PoGo.PokeMobBot.Logic/Routing.cs]: Same as other PR, dont catch generic exceptions.  Catch what you expect, nothing more

File: PoGo.PokeMobBot.Logic/Settings.cs [hunk 3]
```diff
@@ -997,6 +997,8 @@ public class LocationSettings
         public bool RandomizeSpawnOnStart = true;
         public bool UseMapzenApiElevation = false;
         public string MapzenApiElevationKey = "";
+        public bool UseMapzenOptimize = true;
```
Comment (bmdr38) [inline PoGo.PokeMobBot.Logic/Settings.cs]: Default should be false

File: PoGo.PokeMobBot.Logic/Tasks/FarmPokestopsTask.cs [hunk 4]
```diff
@@ -504,6 +508,21 @@ public static class FarmPokestopsTask
...
+        {
+            List<FortCacheItem> list = new List<FortCacheItem>();
+            pokestopList =
+                   pokestopList.OrderBy(
+                       i =>
+                           LocationUtils.CalculateDistanceInMeters(session.Client.CurrentLatitude,
+                               session.Client.CurrentLongitude, i.Latitude, i.Longitude)).ToList();
+            int licount = pokestopList.Count() >= 49 ? 49 : pokestopList.Count();
+
+            list.AddRange(pokestopList.GetRange(0, licount));
```
Comment (bmdr38) [inline PoGo.PokeMobBot.Logic/Tasks/FarmPokestopsTask.cs]: Can just do list.AddRange(pokestopList.Take(49))

**Comment by FabryB** [comment]
There is a "small" problem you didn't consider:

> *optimized_route* 	Returns an optimized route path from point ‘a’ to point ‘n’. Given a list of locations, an optimized route with stops at each intermediate location exactly one time, always starting at the first location in the list and ending at the last location.

This means you are forcing the start to the current position, but you are also forcing the end to the 49th pokestop in list, which is **bad**

**Comment by lorilith** [comment]
Start and end was considered, since end location is likely to be the farthest away, instead of running 49 paths to determine the shortest based on end location, a path from here to the farthest out point seemed to be the best option.  A one to many could be done first to determine which endpoint truly is the farthest, but this would further limit API key usage and limited call capability. (though at the given limit, that would exceed the number of pokestops per day possible).  Would a start->start with 48 locations inside be more desirable?

**Comment by lorilith** [comment]
Errors in routing unhandled would break pathing with current code. not ready for integration

**Pull request closed without merging**